### PR TITLE
[OpenAI] Added 2024-04-01-preview Support. 

### PR DESCRIPTION
**Pull request by mssfang**
resolved: https://github.com/Azure/azure-sdk-for-java/issues/39621

### Features Added

- Added support for service API version, `2024-04-01-preview`. 
- Note that`AOAI` refers to Azure OpenAI.

**Audio**

- Added timestamp granularity to Whisper transcription; this is an array of enumerated string values 
  (word and/or segment) that controls which, if any, timestamp information is emitted to transcription results.
  - `AudioTranscriptionTimestampGranularity` enum to represent the timestamp granularity options for Whisper transcription.
  - `AudioTranscriptionWord` class to represent the word timestamp information in the transcription results.
- Added two new audio format support, 'wav' and 'pcm', to the `SpeechGenerationResponseFormat` enum.

**Chat**

- [AOAI] Added a new property `indirectAttack` in `ContentFilterResultDetailsForPrompt` class to represent the indirect attack results.

### Breaking Changes

- [AOAI] Added a new class `ContentFilterDetailedResults` to represent detailed content filter results, which replaces the
  `customBlocklists` response property's type, `List<ContentFilterBlocklistIdResult>` in 
  `ContentFilterResultDetailsForPrompt` and `ContentFilterResultsForChoice` class.

### CODE REVIEW

File: sdk/openai/azure-ai-openai/src/test/java/com/azure/ai/openai/NonAzureOpenAISyncClientTest.java [hunk 1]
```diff
@@ -396,6 +398,66 @@ public void testGetAudioTranscriptionVtt(HttpClient httpClient, OpenAIServiceVer
...
+
+        getAudioTranscriptionRunnerForNonAzure((modelId, transcriptionOptions) -> {
+            transcriptionOptions
+                    .setResponseFormat(AudioTranscriptionFormat.VERBOSE_JSON)
+                    .setTimestampGranularities(Arrays.asList(AudioTranscriptionTimestampGranularity.SEGMENT));
+
+            AudioTranscription transcription = client.getAudioTranscription(modelId, transcriptionOptions.getFilename(),
+                    transcriptionOptions);
+
+            assertAudioTranscriptionSegments(transcription.getSegments());
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/test/java/com/azure/ai/openai/NonAzureOpenAISyncClientTest.java:434]: should we assert not empty too?

File: sdk/openai/azure-ai-openai/src/test/java/com/azure/ai/openai/models/ChatCompletionsToolCallUnitTests.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+package com.azure.ai.openai.models;
+
+import com.azure.core.util.BinaryData;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.Collections;
+
+/**
+ * Unit tests for ChatCompletionsToolCall.
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/test/java/com/azure/ai/openai/models/ChatCompletionsToolCallUnitTests.java:13]: I would add a brief word in this part of the documentation explaining why these tests are necessary for this particular asymmetry between serialization and deserialization.

**Review by jpalvarezl — APPROVED** [review]
Small optional comments left here and there :) Thanks for the quick implementation 🙌 !

File: .vscode/cspell.json [hunk 3]
```diff
@@ -230,6 +230,7 @@
     "amqp",
     "Amqp",
     "AMQP",
+    "AOAI",
```
[the remark this answers is not in the document]
Comment (mssfang) [reply .vscode/cspell.json:233]: Yeah. I thought about it and decided to not include it until we actually use it. `Lazy v.s. Eager` then took lazy approach.

File: sdk/openai/azure-ai-openai/src/test/java/com/azure/ai/openai/NonAzureOpenAISyncClientTest.java (hunk 1 above)
Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/test/java/com/azure/ai/openai/NonAzureOpenAISyncClientTest.java:434]: we did it inside of [assertAudioTranscriptionSegements](https://github.com/Azure/azure-sdk-for-java/pull/39660/files#diff-98ebf74b657b3f63b0c36d212b741da5adec3261dcc6fa6666a4d482dd6595a6R699) by using azure core util method. 
`assertFalse(CoreUtils.isNullOrEmpty(segments))`

File: sdk/openai/azure-ai-openai/src/test/java/com/azure/ai/openai/NonAzureOpenAIAsyncClientTest.java [hunk 4]
```diff
@@ -399,6 +401,66 @@ public void testGetAudioTranscriptionVerboseJson(HttpClient httpClient, OpenAISe
...
+    @RecordWithoutRequestBody
+    public void testAudioTranscriptionTimestampGranularityInBothSegmentAndWord(HttpClient httpClient, OpenAIServiceVersion serviceVersion) {
+        client = getNonAzureOpenAIAsyncClient(httpClient);
+
+        getAudioTranscriptionRunnerForNonAzure((deploymentName, transcriptionOptions) -> {
+            transcriptionOptions
+                    .setResponseFormat(AudioTranscriptionFormat.VERBOSE_JSON)
+                    .setTimestampGranularities(Arrays.asList(
+                            AudioTranscriptionTimestampGranularity.SEGMENT,
+                            AudioTranscriptionTimestampGranularity.WORD));
```
[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/test/java/com/azure/ai/openai/NonAzureOpenAIAsyncClientTest.java:454]: Yeah. Added more tests.

The error is HttpResponseException with message of asking to correct the response type.

```
com.azure.core.exception.HttpResponseException: Status code 400, "{
  "error": {
    "message": "Timestamp granularities are only supported with response_format=verbose_json",
    "type": "invalid_request_error",
    "param": "timestamp_granularities",
    "code": "timestamp_granularities_format"
  }
}"
```

**Pull request merged**